### PR TITLE
docs(workflow): specify weak-tier acceptance-gate rigor and clean output

### LINKED ISSUE #1550: Specify the weak-tier generation and acceptance-gate contract (rigor + clean-module output)
## Background

Parent roadmap: #1540 (weak-local-field-notes-r2).

`docs/idd-workflow.md`'s [Weak-model guardrails](https://github.com/kurone-kito/idd-skill/blob/main/docs/idd-workflow.md#weak-model-guardrails) section (verified current wording) states a preference for deterministic helpers over prose judgment, but does not specify how rigorous a per-unit acceptance check must be, nor what a generation prompt must demand of the produced artifact. An independent field PoC found both under-specifications cause total failures unrelated to a weak model's actual grasp of the logic: <https://gist.github.com/kurone-kito/8f45ee1ec0c07e8a7fdc8c1f04e83654>.

Evidence: (a) a narrow acceptance check (1-2 cases) let a fragile weak-model generation with a variable-shadowing infinite loop pass undetected; a rigorous acceptance check (multiple boundary cases, explicit assertions, and a timeout) caught it and drove a fix-loop to a robust result — the timeout specifically is what turns an infinite loop into a caught failure instead of a hang. (b) A driver prompt asking for "a complete Python file" caused a weak model to inline the acceptance test (import, asserts, and a print) into the target module; because the module's name matched what the external test imports, the inlined import became a circular self-import, causing an `ImportError` on every retry — a total failure that had nothing to do with the model's understanding of the actual task.

## Proposed change

Add two concrete, checklist-form (not prose-only) items to the "Weak-model guardrails" section:

1. Any per-unit acceptance check gating weak-tier generated output should include multiple boundary cases, explicit assertions, and a timeout — call out the timeout specifically as the mechanism that turns an infinite loop into a caught failure rather than a hang.
2. A generation prompt driving weak-tier output should instruct the model to produce a clean, importable artifact — the target definitions and needed imports only — and explicitly forbid embedding the acceptance test, assertions, prints, or a `__main__`/self-invoking block, and forbid importing the artifact under construction. State that the acceptance check itself must never be shown to the model being tested, since a weak model that sees the test tends to reproduce it inline rather than keep the two separated.

Frame both as guidance for whoever specifies a weak-tier execution contract (an adopter's downstream harness, or the harness-orchestrated execution-mode investigation filed alongside this roadmap), not as a change to this repository's own build or test tooling.

## Acceptance criteria

- [ ] "Weak-model guardrails" states the acceptance-rigor requirement (boundary cases, assertions, timeout) as an explicit checklist item.
- [ ] "Weak-model guardrails" states the clean-artifact generation requirement (no embedded test/assertions/prints/self-invocation, no self-import) as an explicit checklist item, and notes the acceptance check must not be shown to the generating model.
- [ ] Both additions are framed as guidance for a weak-tier execution contract, not as new requirements on this repository's own tooling.
- [ ] `node scripts/audit-docs.mjs --check` passes.

---

_Autopilot suitability: 4 / 5 -- higher is more autopilot-suitable; below the configured floor is human-oriented._

_Effort: S -- author-estimated size; a soft autopilot selection tie-breaker only._

### PR DESCRIPTION
**Pull request by kurone-kito**
# Summary

Adds two checklist-form bullets to "Weak-model guardrails" in `docs/idd-workflow.md` and `idd-template/docs/idd-workflow.md`: acceptance-check rigor (boundary cases, explicit assertions, a timeout) and clean-artifact generation (no embedded test/assertions/prints/self-invocation, no self-import, and never showing the acceptance check itself to the generating model).

## Linked issue

Closes #1550

## Follow-up issues (if any)

- [x] none

## Background / rationale (if material)

Sourced from an independent field PoC (linked from the issue) that found under-specified acceptance-check rigor and generation-prompt requirements caused total failures unrelated to a weak model's actual grasp of the task logic: a narrow (1-2 case) acceptance check let an infinite-loop defect pass undetected, and a driver prompt asking for "a complete Python file" caused a weak model to inline the acceptance test into the target module, creating a circular self-import. Both new bullets are framed as guidance for whoever specifies a weak-tier execution contract, not a change to this repository's own tooling.

A prior-round critique flagged a wording risk: sibling PR #1579's "Full-agent-wrapper degradation" bullet (immediately above, in the following subsection) says to feed the model "only the atomic task and its acceptance context," which reads similarly to this PR's "acceptance check." The new clean-artifact bullet disambiguates explicitly: "the acceptance check itself — the literal test code, assertions, and expected outputs, as distinct from the behavioral acceptance criteria described in the prompt — must never be shown to the generating model."

## IDD impact

- [x] Instruction files changed
- [x] Template files changed
- [ ] Helper scripts changed
- [ ] Config schema changed
- [ ] Security / credential / merge behavior changed

## Verification

- [x] `pnpm lint`
- [ ] `pnpm test` (docs-only change, no test suite impact)
- [x] `node scripts/audit-docs.mjs --check`
- [x] `pnpm run docs:sync:check`
- [x] `node scripts/idd-doctor.mjs` (if applicable)

## Safety

- [x] No secret-bearing examples
- [x] No private repo names / URLs
- [x] Merge policy impact reviewed
- [x] Context budget impact reviewed

## Summary by CodeRabbit

* **Documentation**
  * Added guidance for validating weak-tier generated outputs with boundary cases, explicit assertions, and timeouts.
  * Clarified that generated artifacts must remain clean and importable, without embedded tests, entry points, or self-imports.
  * Documented that acceptance-test code and expected outputs must not be exposed to the generating model.

### CODE REVIEW

File: docs/idd-workflow.md [hunk 1]
```diff
@@ -147,6 +147,23 @@ When a lightweight-tier model runs any part of this loop:
...
+  output should instruct the model to produce a clean, importable
+  artifact — the target definitions and needed imports only — and
+  explicitly forbid embedding the acceptance test, assertions, prints,
+  or a `__main__`/self-invoking block, and forbid importing the
+  artifact under construction. The acceptance check itself — the
+  literal test code, assertions, and expected outputs, as distinct
+  from the behavioral acceptance criteria described in the prompt —
+  must never be shown to the generating model, since a weak model that
+  sees the test tends to reproduce it inline rather than keep the two
+  separated.
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply docs/idd-workflow.md]: **Accepted** — fixed in b91f08ff (gerund form: "rather than keeping the two separated").

File: idd-template/docs/idd-workflow.md [hunk 2]
```diff
@@ -150,6 +150,23 @@ When a lightweight-tier model runs any part of this loop:
...
+  output should instruct the model to produce a clean, importable
+  artifact — the target definitions and needed imports only — and
+  explicitly forbid embedding the acceptance test, assertions, prints,
+  or a `__main__`/self-invoking block, and forbid importing the
+  artifact under construction. The acceptance check itself — the
+  literal test code, assertions, and expected outputs, as distinct
+  from the behavioral acceptance criteria described in the prompt —
+  must never be shown to the generating model, since a weak model that
+  sees the test tends to reproduce it inline rather than keep the two
+  separated.
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply idd-template/docs/idd-workflow.md]: **Accepted** — fixed in b91f08ff (gerund form: "rather than keeping the two separated").

**Comment by kurone-kito** [comment]
_claude-code-a9064f64: review triage snapshot — IDD automation marker. Do not edit._

File: docs/idd-workflow.md [hunk 3]
```diff
@@ -147,6 +147,23 @@ When a lightweight-tier model runs any part of this loop:
...
+  explicit assertions, and a timeout — the timeout specifically is
+  what turns an infinite loop into a caught failure instead of a hang.
+  A narrow check (one or two cases) can let a fragile generation with a
+  real defect pass undetected.
+- **Clean-artifact generation**: a generation prompt driving weak-tier
+  output should instruct the model to produce a clean, importable
+  artifact — the target definitions and needed imports only — and
+  explicitly forbid embedding the acceptance test, assertions, prints,
+  or a `__main__`/self-invoking block, and forbid importing the
+  artifact under construction. The acceptance check itself — the
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply docs/idd-workflow.md]: **Accepted** — fixed in 24ee0c94: generalized to a self-invoking entrypoint block, citing Python's `__main__` guard only as one example.

File: idd-template/docs/idd-workflow.md [hunk 4]
```diff
@@ -150,6 +150,23 @@ When a lightweight-tier model runs any part of this loop:
...
+  explicit assertions, and a timeout — the timeout specifically is
+  what turns an infinite loop into a caught failure instead of a hang.
+  A narrow check (one or two cases) can let a fragile generation with a
+  real defect pass undetected.
+- **Clean-artifact generation**: a generation prompt driving weak-tier
+  output should instruct the model to produce a clean, importable
+  artifact — the target definitions and needed imports only — and
+  explicitly forbid embedding the acceptance test, assertions, prints,
+  or a `__main__`/self-invoking block, and forbid importing the
+  artifact under construction. The acceptance check itself — the
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply idd-template/docs/idd-workflow.md]: **Accepted** — fixed in 24ee0c94: generalized to a self-invoking entrypoint block, citing Python's `__main__` guard only as one example.

**Comment by kurone-kito** [comment]
_claude-code-a9064f64: review triage snapshot — IDD automation marker. Do not edit._

**Comment by kurone-kito** [comment]
_claude-code-a9064f64: review triage snapshot — IDD automation marker. Do not edit._

**Comment by kurone-kito** [comment]
| Field | Value |
| --- | --- |
| Phase | F3 merged |
| Claim | claude-code-a9064f64 / claim-bb4bf5251184 |
| Branch | issue/1550-specify-weak-tier-generation-acceptance |
| Last checked | 2026-07-22T03:36:55.260Z |
| Open blockers | none |
| Next action | F4 cleanup then F5 discover |
| Authoritative by | merge commit 4b002bdb, matched head 7cd35185 |
